### DISCUSSION TITLE
Xcelium simulator and embedded iPython shell

### DISCUSSION THREAD

**Discussion by markusdd** [cocotb/cocotb#4144] [Q&A]
Hi, I was successful in embedding an iPython shell in our testcase so you can basically interactively advance the simulation, even with the waveform window open. Unfortunately the resulting iPyton shell is pretty basic, you got no nice formatting, tab completion etc.

I think this might be due to the log formatter interfering with it.

What is the proper way to get the 'full' iPython experience? Functionally, everything is alright, what I type in does work.

![grafik](https://github.com/user-attachments/assets/fa9a7efb-eb14-4c8e-9627-89de8266cdb6)

**Comment by imphil** [comment]
Have a look at https://github.com/cocotb/cocotb/discussions/3981, which might be what you're after?

**Reply by markusdd** [reply]
Hi, no it isn't.

This is about the ipython embed functionality.

This Jupyter Kernel is nice for playing around but not really suitable for the large stuff we're using.
